executor/oci/hosts.go

```diff
@@ -18,21 +19,21 @@ const hostsContent = `
 ::1	localhost ip6-localhost ip6-loopback
 `
 
-func GetHostsFile(ctx context.Context, stateDir string, extraHosts []executor.HostIP, idmap *idtools.IdentityMapping) (string, func(), error) {
+func GetHostsFile(ctx context.Context, stateDir string, extraHosts []executor.HostIP, idmap *idtools.IdentityMapping, hostname string) (string, func(), error) {
 	if len(extraHosts) == 0 {
 		_, err := g.Do(ctx, stateDir, func(ctx context.Context) (interface{}, error) {
-			_, _, err := makeHostsFile(stateDir, nil, idmap)
+			_, _, err := makeHostsFile(stateDir, nil, idmap, hostname)
 			return nil, err
 		})
 		if err != nil {
 			return "", nil, err
 		}
 		return filepath.Join(stateDir, "hosts"), func() {}, nil
 	}
-	return makeHostsFile(stateDir, extraHosts, idmap)
+	return makeHostsFile(stateDir, extraHosts, idmap, hostname)
 }
 
-func makeHostsFile(stateDir string, extraHosts []executor.HostIP, idmap *idtools.IdentityMapping) (string, func(), error) {
+func makeHostsFile(stateDir string, extraHosts []executor.HostIP, idmap *idtools.IdentityMapping, hostname string) (string, func(), error) {
 	p := filepath.Join(stateDir, "hosts")
 	if len(extraHosts) != 0 {
 		p += "." + identity.NewID()
```

Comment:
this doesn't seem to handle the case where hostname changes for different executions

Reply:
sorry @tonistiigi what do you mean by "different executions", could you share more info? Thanks~

Reply:
There is some weird caching in this function with the `Stat` call. The changes here seem to conflict with that when different hostname values are passed.

Reply:
@tonistiigi This is hard to imagine... From cmdline input's POV, there accepts only one "hostname".  
Do you mean that if there're several cmdline executions with different hostnames, here `makeHostsFile()` may get different hostsnames at the same time? At this case, what should we do to elimite this case? Put the whole `meta` to `GetHostsFile()`?

Reply:
@tonistiigi 
Are you describing case when in Dockerfile hostname can overwritten by setting 
ARG HOSTNAME in several layers?

Reply:
@tonistiigi could you please share more info? your comments will be really helpful to it, many thanks~

Reply:
@tonistiigi could you update us please? if it's not blocker, could you approve this pull request?

Reply:
Different invocations of executor share the same hosts file path and same cache location. A new executor instance is created for every `llb.Exec`.

Reply:
@tonistiigi 
Thanks for your clarification~ 
So adding a `hostname != "buildkitsandbox"` here to create a new hosts file to avoid the confliction?
```go
	if len(extraHosts) != 0 || hostname != "buildkitsandbox" {
		p += "." + identity.NewID()
	}
```

Reply:
Hi @tonistiigi 
So the question how to bind current context with provided hostname ? How to differ between invocations and get passed hostname only for current context?

Reply:
@tonistiigi 
If you can provide code snippet to get the idea, will be good. Looks like this pull requests is hanging too much time and we stuck ...

Reply:
@tonistiigi 
can we drop cache function that reuses context like it done when extrahosts defined?
one solution @jingxiaolu wrote above with additional validation of hardcoded value in if section

other solution is to drop this section at all or these solutions don't solve raised question?
```
if len(extraHosts) == 0 {
_, err := g.Do(ctx, stateDir, func(ctx context.Context) (interface{}, error) {
```

Reply:
@AkihiroSuda @thaJeztah maybe you can help us to find the correct solution? 5 months  PR stuck.

Reply:
@tonistiigi now this hostfile will be created with new one when the special hostname is specified, could you please check whether this is what you expected? Many thanks.

Reply:
@tonistiigi ping

Reply:
HI @tonistiigi 
Could you check the new changes please?

Reply:
Hi @tonistiigi 
I saw that you added this issue to the 0.8.0 milestone. Many thanks!
Could you please review the fix above if it's looking good?